Make webhook scheduled actions unique

thenbrent opened the pull request:
Fixes #22087

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This patch changes how webhooks are scheduled to maintain backward compatibility with a uniqueness requirement present when they were scheduled in WP Cron (with WC < 3.5.0). Specifically, identical webhooks will not be scheduled within 10 minutes of each other.

I've enforced this constraint within `wc_webhook_process_delivery()` instead of in a new method like `add_unique()` in `WC_Queue_Interface` and `WC_Action_Queue` because that would be a breaking change - anything extending `WC_Queue_Interface` would need to implement that new method.

Closes #22087.

### How to test the changes in this Pull Request:

0. Disable Action Scheduler's queue runner by installing [this plugin](https://github.com/Prospress/action-scheduler-disable-default-runner)
1. Create a webhook for order updates
2. Update an order multiple times in the same request, or within concurrent requests
3. View the duplicate webhooks on the **Tools > Scheduled Actions** screen

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### Changelog entry

```
* Fix - Do not schedule duplicate webhooks within 10 minutes of each other to maintain previous behaviour.
```

thenbrent committed:
Make webhooks scheduled actions unique

Fixes #22087

maxrice commented:
@thenbrent and I were chatting about this earlier, and I had a question whether multiple updates to a resource in quick succession (say, an order being updated a few times in the span of a couple seconds) would result in a webhook being sent with stale data. This would've been more of an issue pre-3.5 if only the first webhook in a series of updates was scheduled. However, since the data for a webhook isn't retrieved until the webhooks is sent, this doesn't appear to be an issue.

While out of scope for this PR, it's worth noting that there may be situations where it'd be valuable to receive a webhook with data that reflects the state of the resource at the time the webhook was triggered, rather than when it was sent. FWIW, this is how Shopify webhooks operate, which can be helpful when you need to track how the state of a resource changed over time, but it can be super painful when processing 100k webhooks in short order after an admin bulk updated 10k+ products a few times over. ಠ_ಠ

rrennick commented:
@thenbrent The `600 + gmdate( 'U' )` creates a timestamp 10 minutes in the future so anything pending to run in the next 10 minutes will pass the logic. It should be `$next_scheduled_date->getTimestamp() >= ( 600 + gmdate( 'U' ) )`.

The other bug I found testing this is that the `WC_Action_Queue::get_next()` passes a timestamp to a function expecting a timestring. 

```shell
diff --git a/includes/queue/class-wc-action-queue.php b/includes/queue/class-wc-action-queue.php
index a84fa2d59..4abf6c8af 100644
--- a/includes/queue/class-wc-action-queue.php
+++ b/includes/queue/class-wc-action-queue.php
@@ -125,9 +125,9 @@ class WC_Action_Queue implements WC_Queue_Interface {
        public function get_next( $hook, $args = null, $group = '' ) {
 
                $next_timestamp = as_next_scheduled_action( $hook, $args, $group );

                if ( $next_timestamp ) {
-                       return wc_string_to_datetime( $next_timestamp );
+                       return new WC_DateTime( "@{$next_timestamp}", new DateTimeZone( 'UTC' ) );
                }
 
                return null;
```

thenbrent commented:
Thanks for the review @rrennick!

> anything pending to run in the next 10 minutes will pass the logic. It should be

🤦‍♂️ I've fixed that up (and rebased in the fix).

> The other bug I found testing this

I've proposed a fix for that in #22104 to avoid any issues with this PR holding up its merge.

bryceadams commented:
👍 This will fix an issue impacting a lot of my customers also. Would really recommend it's merged/released this year as it's quite severe for those impacted and causing Woo support requests.

bryceadams commented:
This is a big issue updating every store on 3.5+ that uses webhooks, and the issue gets worse the bigger the store is. I'd really recommend it gets merged and released ASAP.

mikejolley reviewed (approved):
Test successful after applying both patches 👍 Will get this merged for next fix release. Thanks!

bryceadams commented:
@mikejolley thank [you](https://media.giphy.com/media/3zSaOGAwLHhxC/giphy.gif)!

ThomasVeld commented:
I'm still having this issue on API V2, is this only fixed for V3?

rrennick commented:
@ThomasVeld This fix requires that you update WordPress to 5.1 or newer.

ThomasVeld commented:
hi @rrennick, thank you for your quick reply.

I'm on 3.5.2, I've set a webhook for order.created & order.updated, but whenever I update an order I get the response 4 times.
Also with different x-wc-webhook-delivery-id 's, so I can't really build a check if it has been processed already.

rrennick commented:
> I'm on WC 3.5.2 and WP 4.9.9

The function that causes this issue is in core WordPress. This fix takes advantage of a filter added to WordPress 5.1. You need to update to WordPress 5.1 or newer to resolve this issue.

ThomasVeld commented:
Updated to 5.2, still the same issue:

`[2019-05-09 12:33:04] local.INFO: order webhook  
[2019-05-09 12:33:04] local.INFO: 9  
[2019-05-09 12:33:04] local.INFO: order webhook  
[2019-05-09 12:33:04] local.INFO: 9  
[2019-05-09 12:33:05] local.INFO: order webhook  
[2019-05-09 12:33:05] local.INFO: 9  
[2019-05-09 12:33:05] local.INFO: order webhook  
[2019-05-09 12:33:05] local.INFO: 9  
`

rrennick commented:
@ThomasVeld I think you are posting on the wrong issue. This issue is for the slug field in the scheduled action.

mcnaveen commented:
The above-mentioned solution will solve this issue?
"**order.created webhook fires multiple times #28329**"
